### ISSUE TITLE
Do not remove custom element children if there are no inner components

### ISSUE THREAD

**Issue by bob2517** [Active-CSS/active-css#49]
Reactive host attribute variables are not working when they are set as direct children of a custom element. This may be just an issue with 2.4.0 as children directly under a custom element hadn't been documented or even thought about. But in 2.4.0, children can be passed over to an inner component, so direct children of custom elements without an inner component has become a thing. Host attribute variables work fine from inside inner components as per the docs.

The good news is that as direct children, reactive variables work under custom elements. But reactive host attribute variables do not. I'm fixing this next. I'll check the other variable types while I'm at it.

**Comment by bob2517** [comment]
Back on this now. I needed to re-acquaint myself with Visual Foxpro for work this last week, hence the absence. Should push a fix onto the branch today or tomorrow.

**Comment by bob2517** [comment]
Weirdly, there is some updating of host elements going on, that actually looks fine. Outer element vars look fine, inner div attributes look fine. There's no scoping on the host element, but that's to be expected with direct children when there are no additional inner components. But text within the innermost div that I'm testing is disappearing. So it appears that there's only a problem with updating inner text nodes in direct children that reference host variables. At least it's an isolated issue and not a problem with the general method. So it should be a localised fix.

**Comment by bob2517** [comment]
I think I've found out what is going on. The introduction of the passing of child nodes into custom elements and processing the host variables prior to handling components wipes out all inner text nodes. That would do it every time. Sorting out the fix now...

**Comment by bob2517** [comment]
Thought I had it sussed but not quite. Just need to work out how to handle nested child descendants with separate hosts and that should be that. It's a completely different scenario in comparison to handling components so at least it's isolated in the code.

**Comment by bob2517** [comment]
Note to self: The current handling for non-component hosted divs is outside of the specific component rendering inside the actual render command, which was fine for my simple example, but of course it looks like it also needs to be incorporated into the components themselves in case they include non-component hosting divs. So remove the code from the render command and have it called from within the component rendering as well as the non-component top-level rendering. The code currently only works on only one level of host - it doesn't work with nested hosts, so it's going to need a recursive drilldown of hosts for initial rendering. Something like that.

Once the initial rendering is done the reactivity should just work. It's just the initial render that needs sorting out.

Get a really complex test case set up before sorting out the fix so everything's covered.

**Comment by bob2517** [comment]
Just to clarify for myself what is going one, I'm listing out the rules regardless @host variables.

A component here is defined as a {|myComponent} reference in the code.

The rules for handling @host references are:
- a) {@host} works for component hosts, or custom elements that have children but don't necessarily have an inner component.
- b) Custom elements can also use {{@host}} to have reactive attributes to follow native rules.

There are 3 handlings in the core.
1. Component hosts: Regardless of whether a custom element is used or not, the component host will act as the host for host variable referencing.
2. Custom element host rule 1:
If the custom element has an inner component, then when rendering the child gets wiped and needs to be referenced inside the inner component somewhere, otherwise it isn't going to render. This allows custom child element passing in to the component like React.
3. Custom element host rule 2:
If there is no inner component and there are @host attributes being referenced inside the custom element, the child does not get wiped. The custom element becomes a non-private component host.

Points 1 and 2 are working in the core. Point 3 was never clearly defined and built in, hence that is the actual issue. It's more like a feature missing than a bug. I tried to write something that should have instinctively worked and the core wasn't set up to handle it.

**Comment by bob2517** [comment]
The more interesting part of fixing this is that point 3 needs to work with nested custom elements. The good news is that at any one time during a nested component render, the string getting prepared for rendering will contain the full nest needed for working out if the host has any inner components or not. The trick is to distinguish rule 2 from rule1 so the child of the component doesn't get removed and the host is correctly set up.

**Comment by bob2517** [comment]
A bit more to point 3. Because of the dynamic nature of CSS, it is very possible that the developer may want to render reference to a @host variable after initial render. Hosts are found dynamically at render time anyway, so that's no biggie. But the custom element needs to be set up as a host at initial render for speed later on.

It looks like the simplest thing is to auto-define all custom elements that have no inner component as non-private hosts.

Look into this as the code handling: On rendering, check if element is a custom element. If it has no inner component about to render, set it as a non-private component host. It's possible that everything else will take care of itself and then we can remove the existing tree traversal and speed things up a bit.

**Comment by bob2517** [comment]
Sort-of works offline. Doesn't yet work with nested custom elements that share the same attributes with different nested values. Bit of a niche scenario, like why would anyone want to do that? They could just use different attribute names and it would be fine.

I could set a rule that nested custom elements with no inner components must have different attribute names. That would allow attributes to pass down through components, which might be useful. Meh. The only thing is that it would mean a different scoping rule and make it more complicated. Dunno. I could check-in what's there now and it would probably be ok. I'm going to run some more tests before changing anything else.

**Comment by bob2517** [comment]
One thing I need to do though - there's scope (no pun intended) in the render command for it to be a bit faster now that what was there regarding handling host attributes has been removed from the mix, so it needs re-factoring before anything is committed.

**Comment by bob2517** [comment]
I _think_ this is almost done now. Just 3 extra lines of code to sort it all out with nesting, etc. after removing an unnecessary section of code. Just one concept example is breaking which needs fixing and I want to re-factor something which may or may not give a slight performance improvement - it's a tough one to call as I had to add a new treewalker section to sort out this issue, so I dunno - may be swings and roundabouts on that. I've also trimmed observableslim down a bit to reduce the core size slightly too for minification. Pushing for a commit this evening - if not tonight it will definitely be tomorrow.

So when this issue is sorted out finally, nested custom elements with no-components should have the correct host, and custom elements will have this same behaviour when rendered within components themselves. Host attributes are strictly contained to the custom element container or the component host, and will _not_ bubble down the DOM.

If anyone wants to report anything strange or not working in the future then that would be helpful. It's literally just me finding all the bugs through trying out scenarios. I've had zero bug reports from anyone else since forever. I dunno if that's a good thing or not...
